### ISSUE TITLE
Task based local spawning

### ISSUE THREAD

**Issue by cyberblast** [ScreepsOCS/screeps.behaviour-action-pattern#343]
_From @cyberblast on December 20, 2016 20:42_

Actually we now have 2 concurring spawning systems. Which is not good. 
We should try to get rid of the old one. 

To do that, we need a big task, to manage local (owned rooms) population. 
We will need different logic per RCL, we could maybe use a master task and different sub tasks. 
There should also be a different handling for "first" and "later" rooms (pioneers, worker numbers etc...).

_Copied from original issue: cyberblast/screeps.ocs.internal#116_

**Comment by karlthepagan** [comment]
I think we need to broach the subject of this rearchitecture.

We need more guidance as to how tasks fit into the behaviour-action-pattern.

I also feel that some per-RCL logic can be resolved when #453 is merged.

**Comment by karlthepagan** [comment]
Also see comments in #445 - we need clear guidance on when tasks can spawn, and how these spawn and assignment decisions should have inner cohesion vs being able to resolve conflicts with other modules.

**Comment by karlthepagan** [comment]
My initial vision: do not carry out actions directly from tasks. Assign behaviours and tasks or create temporary commands which all behaviours will respect.

For example a "danger" command will tell remote miners to retreat to another room when an invasion happens. This danger flag can be set by the mining task, but the miner actions should be very simple and isolated to the mining behaviour.

**Comment by karlthepagan** [comment]
The same "danger" command could be re-used to have different initial requirements based on the tasks. I know this is a slight overlap of the Room.invasion getter, but the intent is for this to be more specific (and a task-based reaction) than the generic invasion.

**Comment by karlthepagan** [comment]
I guess the biggest conflict at this point is how spawning happens. Maybe that's the extent of my angst.

**Comment by whardeman** [comment]
@karlthepagan Sounds like we need to resolve that question across the board and then derive from that decision all other architectural decisions?

**Comment by karlthepagan** [comment]
It's almost like we should have behaviors for structures... Might be a can of worms.

**Comment by whardeman** [comment]
@karlthepagan So that comes back to an idea I had a while ago... a room manager would make these kinds of decisions about spawning based on the current mode/state of a room. The default state would balance priorities between energy and defense. A room state could be overridden manually, say, if I want to designate a room (deep in the interior of my empire) as purely mining oriented vs a frontier room that will be producing lots of military creeps for conquest. 

I don't want spiraling complexity, so this idea may be crap. But I thought developing a set of logic that was primarily dedicated to managing a specific room, rather than having tasks fight it out in the spawn queues might work. I think this could ultimately be expanded to an empire level where whole clusters of rooms are managed in this way, but that's getting way too ahead of itself.

**Comment by cyberblast** [comment]
> I guess the biggest conflict at this point is how spawning happens. 
> we need clear guidance on when tasks can spawn, and how these spawn and assignment decisions should have inner cohesion vs being able to resolve conflicts with other modules

1. I think we should add more priorities (more queues or a rating system) for task based spawning. Or add the possibility to define rules for task based queued spawn jobs. To allow better control of which creep may get spawned when. 
2. You should use possibilities to evaluate the environment within the task before simply adding a creep to the queue. (e.g. don't queue a reserve creep, when controller in target room has > 2500 reserve points)
3. Split it up into atomic tasks, which may trigger other events them self. Setting values via tasks is a bad idea, because you'll never know in which order they get called. That's why we'll need to do an environment analysis first, then trigger events.

**Comment by Dasku** [comment]
Ok this problem is starting to make more sense to me now and I'm starting to see it in action.

I have a RCL 6 room (and I think this specific issue only happens to some rooms at RCL 6 only, will test 7 when I get a RCL7 room later), it has access to 4 2 source rooms, it cannot make privateers fast enough, the spawn is never free, and therefore as reserve task uses the Low spawn queue its never getting a chance to produce the claimers to reserve the exploit rooms.

**Comment by cyberblast** [comment]
> it cannot make privateers fast enough, the spawn is never free, and therefore as reserve task uses the Low spawn queue its never getting a chance to produce the claimers to reserve the exploit rooms. 

as we are spawning a reserver for each exploit flag, we could implement a rule to only queue privateers for reserved rooms. This way we enforce spawning a claimer once reservation runs out.

**Comment by Dasku** [comment]
Sounds like a good idea, or alternatively (so doesnt stop spawning miners or haulers for a while until the room is reserved again) use a High prio claimer if the room is below 500 ticks

**Comment by cyberblast** [comment]
no. I'd not consider that as high. 
High should be real critical stuff. like defenders or local home economy. 
but Medium would be fine. privateers could go by low...

**Comment by karlthepagan** [comment]
Automatically putting anything on the high priority queue can lead to resource starvation. Either we need a setup queue higher priority than high (for in-room recovery operations) or we should move everything out of the high queue into medium until the spawn refactor.

**Comment by chobobobo** [comment]
@karlthepagan is the branch change/setupQueue meant to address this issue?

**Comment by karlthepagan** [comment]
After having learned this game and the code base a lot more I'm convinced that task-based spawning should be the norm.

I think that worker is the only unit which needs to be exceptional in it's timing and when it spawns on demand. A 200 energy worker should be queued as high priority (or maybe 90% priority?) to resolve dying rooms.

I don't see why anything else should be flexibly spawned based on current available energy.

**Comment by karlthepagan** [comment]
These are the reasons the setupQueue must die:
* the job a creep is capable of changes depending on its size
* the job a creep is assigned to should be scheduled well in advance of it spawning
* the urgency of a spawning directly relates to when creeps need their jobs replaced

setup queue has too many moving parts:
* creep body size is arbitrary until the API is invoked
* it is the cause of priority inversion in the scheduled queues
* this means that a busy room has empty extensions and immediately gets worse

after 5 respawns... yeah it makes a lot of sense

**Comment by chobobobo** [comment]
Would it perhaps work if spawns used event callbacks to poll receivers of spawns eg. flags/rooms and they reply with the information you need to determine the best next spawn? That way we know a spawn is not being started that is not currently needed, and it would allow the receivers to be more deterministic about what is being spawned, and when to expect it.  The spawn could then make the decision based on a score assigned, with distance being a primary factor.  Is this in alignment with eliminating setupQueues, or did you have something else in mind?

**Comment by karlthepagan** [comment]
Yes, I think that we could try to incorporate callbacks into the spawn queue where any features of the setup style queue are needed (flexible body size).

**Comment by karlthepagan** [comment]
IMO 0.7.1 is the time to close this out finally.

Items needed to complete this refactor:
* adding validation and on-demand spawning into task.js (complete?)
* integrate predictedRenewal into task.js validation
* a callback which will allow a worker or hauler to spawn with less than the requested body size as a collapse safety measure
* spawn queue visualization?
